ml-agents/mlagents/trainers/reward_providers/curiosity_reward_provider.py

```diff
@@ -0,0 +1,260 @@
+
+
+def swish(x):
+    """
+    TODO : MOVE SOMEWHERE BETTER
+    """
+    return x * torch.sigmoid(x)
+
+
+class Swish(torch.nn.Module):
+    """
+    TODO : MOVE SOMEWHERE BETTER
+    """
+
+    def forward(self, data: torch.Tensor) -> torch.Tensor:
+        return torch.mul(data, torch.sigmoid(data))
+
+
+def actions_to_onehot(
+    discrete_actions: torch.Tensor, action_size: List[torch.Tensor]
+) -> List[torch.Tensor]:
+    """
+    Splits a discrete action Tensor (of integers) into its one hot representations.
+    Returns a list of Tensors (One Tensor per branch)
+    :param discrete_actions: A Tensor of discrete actions.
+    :param action_size: List of ints containing the number of possible actions for each branch.
+    :return: A list of one hot Tensors (one or each branch).
+    """
+    onehot_branches = [
+        torch.nn.functional.one_hot(_act.T, action_size[i]).float()
+        for i, _act in enumerate(discrete_actions.long().T)
+    ]
+    return onehot_branches
+
+
+def break_into_branches(
+    concatenated_logits: torch.Tensor, action_size: List[torch.Tensor]
+) -> List[torch.Tensor]:
+    """
+    Takes a concatenated set of logits that represent multiple discrete action branches
+    and breaks it up into one Tensor per branch.
+    :param concatenated_logits: Tensor that represents the concatenated action branches
+    :param action_size: List of ints containing the number of possible actions for each branch.
+    :return: A List of Tensors containing one tensor per branch.
+    """
+    action_idx = [0] + list(np.cumsum(action_size))
+    branched_logits = [
+        concatenated_logits[:, action_idx[i] : action_idx[i + 1]]
+        for i in range(len(action_size))
+    ]
+    return branched_logits
+
+
+def dynamic_partition(
+    data: torch.Tensor, partitions: torch.Tensor, num_partitions: int
+) -> List[torch.Tensor]:
+    """
+    Torch implementation of dynamic_partition :
+    https://www.tensorflow.org/api_docs/python/tf/dynamic_partition
+    Splits the data Tensor input into num_partitions Tensors according to the indices in
+    partitions.
+    :param data: The Tensor data that will be split into partitions.
+    :param partitions: An indices tensor that determines in which partition each element
+    of data will be in.
+    :param num_partitions: The number of partitions to output. Corresponds to the
+    maximum possible index in the partitions argument.
+    :return: A list of Tensor partitions (Their indices correspond to their partition index).
+    """
+    res: List[torch.Tensor] = []
+    for i in range(num_partitions):
+        res += [data[(partitions == i).nonzero().squeeze(1)]]
+    return res
+
+
+class CuriosityRewardProvider(BaseRewardProvider):
+    def __init__(self, specs: BehaviorSpec, settings: CuriositySettings) -> None:
+        super().__init__(specs, settings)
+        self._network = CuriosityNetwork(specs, settings)
+        params = list(self._network.parameters())
```

Comment:
You can just pass `self._network.parameters()` directly to the optimizer. No need to turn it into a list.

Reply:
Nice, thanks